### PR TITLE
Daemon controller Implementation

### PR DESCRIPTION
**Pull request by vishh**
Most of the work was done by @AnanyaKumar based on his original proposal. I have rebased his work to current HEAD.

cc @bgrant0607 @kubernetes/goog-ux @kubernetes/goog-testing @kubernetes/goog-control-plane

### CODE REVIEW

File: test/e2e/daemon.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	"k8s.io/kubernetes/pkg/util/wait"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Daemon", func() {
+	var c *client.Client
+
+	BeforeEach(func() {
```
Comment (quinton-hoole) [inline test/e2e/daemon.go]: Please use the standard kubernetes/test/e2e/framework.go for your client and run your tests in their own namespace (not namespace.Default).

File: test/e2e/daemon.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+		daemonReaper, err := kubectl.ReaperFor("Daemon", c, nil)
+		Expect(err).NotTo(HaveOccurred())
+		_, err = daemonReaper.Stop(ns, simpleDaemon.Name, 0, nil)
+		Expect(err).NotTo(HaveOccurred())
+		err = wait.Poll(retryInterval, retryTimeout, checkRunningOnNoNodes(c, label))
+		Expect(err).NotTo(HaveOccurred())
+	}()
+
+	By("Check that daemon pods launch on every node of the cluster.")
+	Expect(err).NotTo(HaveOccurred())
```
Comment (quinton-hoole) [inline test/e2e/daemon.go]: Duplicate?  I don't think that you need this line?

File: test/e2e/daemon.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+		}
+		return checkDaemonPodOnNodes(c, selector, nodeNames)()
+	}
+}
+
+func checkRunningOnNoNodes(c *client.Client, selector map[string]string) func() (bool, error) {
+	return checkDaemonPodOnNodes(c, selector, make([]string, 0))
+}
+
+func testDaemons(c *client.Client) {
```
Comment (quinton-hoole) [inline test/e2e/daemon.go]: Please break up this function, it's over 100 lines long.

**Comment by quinton-hoole** [comment]
e2e tests are failing - you'll need to fix that.
Other than the few comments I made, test/e2e/daemon.go code look fine.
I have not reviewed the other code.

File: test/e2e/daemon.go (hunk 1 above)
Comment (vishh) [inline test/e2e/daemon.go]: Done.

File: test/e2e/daemon.go (hunk 3 above)
Comment (vishh) [inline test/e2e/daemon.go]: it containers multiple subtests. I notice other tests having multiple sub-tests inline as well. The other option is to define one function per test.

File: test/e2e/daemon.go (hunk 2 above)
Comment (vishh) [inline test/e2e/daemon.go]: Good catch! Fixed.

**Comment by vishh** [comment]
Thanks for the quick review @quinton-hoole. Still working on stabilizing the patches.

**Comment by AnanyaKumar** [comment]
Which e2e tests are failing? (I don't have access to Jenkins)

**Comment by AnanyaKumar** [comment]
@googlebot I'm ok with with these commits being contributed to Google.

File: test/e2e/daemon.go (hunk 3 above)
Comment (AnanyaKumar) [inline test/e2e/daemon.go]: I agree that the function can be broken up :)

However, I think we should not break it up into 2 disjoint tests. We should test that 2 daemon controllers can co-exist together (as long as their selectors are different).

**Comment by quinton-hoole** [comment]
Hi @AnanyaKumar 
Thanks for the contribution!  The tests that failed are listed below.  You should be able to repro those errors locally as described in https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#flag-options

```
[FAILED] Kubernetes e2e suite.Kubectl client Kubectl describe should check if kubectl describe prints relevant information for rc and pods (from (junit_11.xml))

Stacktrace

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:323
Aug 24 18:01:13.717: Error running &{/jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@2/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://104.197.85.179 --kubeconfig=/var/lib/jenkins/jobs/kubernetes-pull-build-test-e2e-gce/workspace@2/.kube/config describe pod redis-master-2ia09 --namespace=e2e-tests-kubectl-e3hf9] []  <nil>  error getting daemons: the server could not find the requested resource
 [] <nil> 0xc2087d7a00 exit status 1 <nil> true [0xc208051d48 0xc208051d68 0xc208051d88] [0xc208051d48 0xc208051d68 0xc208051d88] [0xc208051d60 0xc208051d80] [0x6890a0 0x6890a0] 0xc2085e3500}:
Command stdout:

stderr:
error getting daemons: the server could not find the requested resource

[FAILED] Kubernetes e2e suite.Daemon should launch a daemon pod on every node of the cluster (from (junit_13.xml))

Stacktrace

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/daemon.go:45
Expected error:
    <*errors.StatusError | 0xc208291d00>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {SelfLink: "", ResourceVersion: ""},
            Status: "Failure",
            Message: "node \"pull-e2e-1-minion-1b6j\" cannot be updated: the object has been modified; please apply your changes to the latest version and try again",
            Reason: "Conflict",
            Details: {
                Name: "pull-e2e-1-minion-1b6j",
                Kind: "node",
                Causes: nil,
                RetryAfterSeconds: 0,
            },
            Code: 409,
        },
    }
    node "pull-e2e-1-minion-1b6j" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
not to have occurred

**Comment by AnanyaKumar** [comment]
Oh yeah, I'm not sure what I was thinking.

The e2e tests will fail. I added daemon to unsupported (in the API) as discussed with @bgrant0607, which temporarily disables daemon functionality.  We need to add daemon to the experimental API. We didn't do this earlier because experimental API was still being worked on. 

I'm not sure how to proceed with the PR. If you remove daemon from unsupported, and add 2 lines to pkg/master/master.go, the tests should pass. If the experimental API is ready, perhaps it's time to move daemon to the experimental API.

**Comment by bgrant0607** [comment]
Yes, please move Daemon to pkg/expapi.

**Comment by bgrant0607** [comment]
And please move the kubectl changes to a separate PR.

Also, eventually, status should be written using a /status subresource.

**Comment by vishh** [comment]
@AnanyaKumar: Do you plan to work on the changes mentioned here?

**Comment by AnanyaKumar** [comment]
@bgrant0607 thanks for the response! @vishh Sorry for not responding, I was offline today. Thanks @mikedanese for sending the PRs to follow up on this.

**Pull request closed without merging**